### PR TITLE
Update shell lecture notes + add advanced shell

### PR DESCRIPTION
**Pull request by bluka479**
Update beginner and intermediate shell notes as required (moving ssh from intermediate from beginner, removing git from intermediate, add package manager notes), and create advanced shell lecture notes.

### CODE REVIEW

**Comment by bluka479** [comment]
Should be the last of the content updates for shell lectures, excluding typos and errors.

File: _2024/shell-intermediate.md [hunk 1]
```diff
@@ -164,6 +173,31 @@ The access of a file can be changed with `chmod`. Again, this can only be done b
 
 Both of the above can also be applied recursively to change the permissions of an entire directory structure. Do this with caution! For details on how to use these commands, take a look at their manual pages.
 
+# Package managers
+
+The primary method of installing packages for linux is usually using a package manager from the command line. There is some variation of package manager based on linux distribution, but most use one of the following: `apt` (Debian-based), `pacman` (Arch-based), `dnf` (RedHat-based), or `zypper` (SUSE-based). For simplicity's sake, we'll be demonstrating how to use `apt`, as it is the most popular package manager, and the principles should carry on to the other package managers through different command names.
```
Comment (LMBishop) [inline _2024/shell-intermediate.md:178]:
```suggestion
The primary method of installing applications on Linux systems is by using a package manager. Each Linux distribution has its own package manager, but most use one of the following: `apt` (Debian-based), `pacman` (Arch-based), `dnf` (RedHat-based), or `zypper` (SUSE-based). For simplicity's sake, we'll be demonstrating how to use `apt`, as it is the most popular package manager, and the principles should carry on to the other package managers through different command names.
```

File: _2024/shell-intermediate.md [hunk 2]
```diff
@@ -164,6 +173,31 @@ The access of a file can be changed with `chmod`. Again, this can only be done b
...
+
+Since package managers edit package files system-wide, root privileges are needed to install, delete, and update package repositories, so most commands will be prefixed by `sudo`.
+
+**Note: Take the same care with using `sudo` now as you always should.**
+
+As as example, let's install `cowsay`:
+
+First, try running `cowsay "Hello!"`. Usually, the `cowsay` command isn't installed, so `bash` should return a `comand not found` error.
+Let's start by updating our package repositories with `sudo apt update`. Run `sudo apt install cowsay` and enter your password to install the package. Folloing this, run `cowsay "Hello!"` and notice that the command is now available. To remove `cowsay`, run `sudo apt remove cowsay`.
```
Comment (LMBishop) [inline _2024/shell-intermediate.md:187]: Maybe it's better for paragraphs involving a lot of separate commands to be broken up using code blocks with output examples instead? E.g.

```bash
sudo apt update
# [...]

sudo apt install cowsay
# [sudo] password for user:

cowsay "Hello"
#  _______
# < Hello >
# -------
#        \   ^__^
#         \  (oo)\_______
#            (__)\       )\/\
#                ||----w |
#                ||     ||
```

File: _2024/shell-advanced.md [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+
+You might have noticed that piping to `grep` doesn't filter error messages. This is because `grep` only filters STDOUT. This means that error messages that get sent through STDERR don't get processed by `grep`, and we need to redirect our standard streams to fix this.
+
+If we move the output from STDERR to STDOUT, `grep` will be able to filter it.
+We can do this with the syntax `2>&1`. This moves data from STDERR (fd `2`) to STDOUT (fd `1`), using an `&` before `1` to avoid ambiguity between writing to a file named `1`.
+
+Let's practice this by running `python3 -c "raise Exception('This goes to STDERR')"`. This runs the python code `raise Exception('This goes to STDERR')`, which throws an error and outputs context data around it. Trying to filter the output with `python3 -c "raise Exception('This goes to STDERR')" | grep STDERR` doesn't work, since the python error output goes to STDERR, but running `python3 -c "raise Exception('This goes to STDERR')" 2>&1 | grep STDERR` does, since it moves the STDERR stream to STDOUT, which `grep` does filter. As well as this, STDOUT can be redirected to STDERR to properly log errors, such as with `echo 'error here!' >&2`.
```
Comment (LMBishop) [inline _2024/shell-advanced.md:46]: Same as my other comment in `shell-intermediate.md` about breaking up large paragraphs with lots of code with examples of output.

**Review by LMBishop — COMMENTED** [review]
Looks good, I like the topics covered, just got a few minor suggestions & typo fixes. I don't think `tinky-winky` exists anymore though so that section might need to be rewritten.

File: _2024/course-shell.md [hunk 4]
```diff
@@ -233,98 +159,72 @@ missing:~$ ls -l /home
...
...g any work, we should then connect to a lab machine, which we can do with the command `ssh-lab`. We then get asked for our password again to log on to the machine. Once logged in, we have the same shell as if we sat down at the lab machine physically. Typing out a long password every time you wish to login can become cumbersome (especially so if you are using tools that connect on your behalf). To resolve this, we can use an SSH key. An SSH key consists of two parts: a private key and public key. In short, the public key is used by others to encrypt data to be sent to you, and the private is used by you to decrypt data sent to you. Since the private key serves as your identity, the key must be kept secret, ideally with a password used to encrypt it. For the university systems, password based login is always available so a key is not strictly required. Most servers however are configured with password based login disabled and only allow key login. This is due to the low security of the average password, possibility of brute forcing them, and the possibility of an attacker stealing the password with a malicious server. For these security reasons key authenticated is widely preferred.
```
[the remark this answers is not in the document]
Comment (bluka479) [reply _2024/course-shell.md:198]: does cs currently have another bastion server name?

File: _2024/shell-advanced.md [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+**Note:** `disown` can also be used with a specific process ID, for example `disown 245` to remove a specific process from the shell job control list.
+
+
+# Multiplexing
+
+Often when using the shell, you might want to see and access multiple shell processes in one tab. To do this, we can use terminal multiplexing. Mutiplexing allows a user to split a terminal into multiple sections, save and restore sessions, and edit theme colouring. One of the most popular terminal multiplexers is `tmux`, and we'll be going through the basics of using `tmux` here.
+
+To start, install the `tmux` package.
+
+**Note:** To always automatically start tmux when you open your terminal, run `echo -e '\nif command -v tmux &> /dev/null && [ -n "$PS1" ] && [[ ! "$TERM" =~ screen ]] && [[ ! "$TERM" =~ tmux ]] && [ -z "$TMUX" ]; then\n  exec tmux\nfi' >> ~/.bashrc`. To disable this, delete the 3 lines starting with `if command -v tmux`.
```
[the remark this answers is not in the document]
Comment (bluka479) [reply _2024/shell-advanced.md:80]: probably a good idea to briefly explain the command, but .bashrc was covered in shell intermediate i think. is it worth mentioning it again here?

**Comment by bluka479** [comment]
Your review's much appreciated thanks, I'll go over the comments soon and correct them.

**Comment by bluka479** [comment]
Checking if this alright to merge now?

**Comment by mariusmue** [comment]
Thanks! Merging this in for now - the changes for session 1 look all good, and changes for later sessions can be addressed when the time comes.

**Pull request merged**